### PR TITLE
feat(quotas): Introduce scoped quotas

### PR DESCRIPTION
**Pull request by jan-auer**
A large-scale refactor of quotas and rate limits.

## Quotas

This PR introduces _scoped quotas_ (see https://github.com/getsentry/sentry/pull/17248) for the changes. Quotas now specify a set of `DataCategories` and `QuotaScope` and are applied selectively based on the ingested item. 

In the project configuration, quotas have moved from individual `publicKeys` into the top-level `config` attribute. Organization- and project-scoped quotas are no longer repeated for each keys. The former configs are shimmed by `LegacyQuota` and converted with the following rules:

- `prefix` is moved into `id`.
- `subscope` is moved into `scope_id`.
- The `scope` is always `Organization`, except for the quotas with id `"k"` (`Key`) and `"p"` (`Project`). These were the only ones with a different scope. 
- Categories are all categories except for `Attachment` and `Session`, since these are not allowed to be ingested by legacy Sentry. Also, `Transaction` will soon be removed from this list in favor of explicit flags once Sentry has been updated.

## Item Scoping

The common interface to convey scoping information is `relay_quotas::ItemScoping`. It contains identifiers of all scopes, as well as the data category. Both quotas and rate limits use item scoping to determine whether an item is to be dropped.

The work on item scoping is not finished, as proper inference of `DataCategory` is not implemented and scopes are not passed around properly in `EventManager`.

## Rate Limits

Rate limits are now scoped as well. There are the following rate limit scopes:

 - `RateLimitScope::Organization`: based on the organization id
 - `RateLimitScope::Project`: based on the project id
 - `RateLimitScope::Key`: based on the actual public key. In contrast, `Quota` uses the key id. This is designed differently because external Relays and the fast-path always have access to the public key reliably in the headers, but not necessarily the key id. 

The rate limit scope can be converted from `QuotaScope` and `ItemScoping`.

All internal APIs now pass around a `RateLimits` object, which is a collection of rate limits per scope and categories. These objects can be merged and reside on the `Project` actor for checking in the fast path. Also, `UpstreamRelay` returns rate limits by parsing the `X-Sentry-Rate-Limits` header which can be merged into the project.

## Code Reorganization

The following components have moved:

 - We had duplicated logic handling unix timestamps. There is now a `UnixTimestamp` type in `relay-common`.
 - Everything related to quotas and rate limiting has moved into the `relay-quotas` crate. It contains a feature to conditionally compile the redis rate limiter. This particularly also involves the `RetryAfter` type which can be parsed from a header. 
 - Redis has been factored out into the `relay-redis` crate. The interface now wraps the internals of whether it's a cluster or single client and implements connections to abstract over it. It still has the _noop_ implementation for convenience, for the time being.
 - Configuration structs for Redis has been moved to `relay-redis`, which `relay-config` now depends on. This localizes the configuration and makes the dependency graph a little nicer.

### CODE REVIEW

**Comment by jan-auer** [comment]
This still needs correct quota scope handling internally, then we can merge it. The follow-up will be to ensure that we block unsupported event types. The first version will assume `DataCategory::Error` for all events and not handle attachments.

**Comment by jan-auer** [comment]
This PR has a few limitations, but I think we should merge it regardless now:
- No inference of data categories. All events are assumed as errors. 
- No shared rate limits across projects from the same organization.
- Not rate limiting for sessions or attachments yet.
- Duplication of the `ItemScope` logic.

File: relay-quotas/src/rate_limiter.rs [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+                invocation.arg(quota.limit());
+
+                invocation.key(refund_key);
+                invocation.arg(quota.expiry().as_secs());
+
+                tracked_quotas.push(quota);
+            } else {
+                // This quota is neither a static reject-all, nor can it be tracked in Redis due to
+                // missing fields. We're skipping this for forward-compatibility.
+                log::warn!("skipping unsupported quota");
```
Comment (untitaker) [inline relay-quotas/src/rate_limiter.rs]: can you push some extra data into sentry-sdk scope or put them into the log msg? Depending on what would group better

File: relay-server/src/actors/upstream.rs [hunk 2]
```diff
@@ -133,16 +190,20 @@ impl UpstreamRelay {
...
+                    let retry_after = headers
                         .get(header::RETRY_AFTER)
-                        .and_then(|v| v.to_str().ok())
-                        .and_then(|s| s.parse::<f64>().ok())
-                        .unwrap_or(0f64)
-                        .max(0f64)
-                        .ceil();
+                        .and_then(|v| v.to_str().ok());
+
+                    let rate_limits = headers
```
Comment (untitaker) [inline relay-server/src/actors/upstream.rs:203]: Make sure you can deal with multiple headers here, as proxies are allowed to split `Foo: a,b` into `Foo: a\nFoo: b`

**Review by untitaker — APPROVED** [review]

**Review by untitaker — APPROVED** [review]
Jan walked me through the changes. I think this looks fine at a high level but it's generally too large to review line-by-line.

File: relay-server/src/actors/upstream.rs (hunk 2 above)
Comment (jan-auer) [reply relay-server/src/actors/upstream.rs:203]: Excellent point, this will require to collect the strings. Thanks!

Comment (jan-auer) [reply relay-server/src/actors/upstream.rs:203]: Alright, the fix in https://github.com/getsentry/relay/pull/503/commits/331df576f28fc28d789806ef0c34e608194fbb13 should do.

**Pull request merged**